core-bundle/tests/Controller/AbstractBackendControllerTest.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Tests\Controller;
+
+use Contao\CoreBundle\Controller\AbstractBackendController;
+use Contao\CoreBundle\Security\Authentication\Token\TokenChecker;
+use Contao\CoreBundle\Tests\TestCase;
+use Contao\Session;
+use Contao\System;
+use Doctrine\DBAL\Driver\Connection;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+use Twig\Environment;
+
+class AbstractBackendControllerTest extends TestCase
+{
+    public function testAddsAndMergesBackendContext(): void
+    {
+        $controller = new class() extends AbstractBackendController {
+            public function fooAction(): Response
+            {
+                return $this->render('custom_be.html.twig', ['foo' => 'bar', 'version' => 'my version']);
+            }
+        };
+
+        // Legacy setup
+        \Contao\Environment::reset();
+        (new Filesystem())->mkdir($this->getTempDir().'/languages/en');
+
+        $GLOBALS['TL_LANG']['MSC'] = [
+            'version' => 'version',
+            'dashboard' => 'dashboard',
+            'home' => 'home',
+            'learnMore' => 'learn more',
+        ];
+
+        $GLOBALS['TL_LANGUAGE'] = 'en';
+
+        $_SERVER['HTTP_USER_AGENT'] = 'Contao/Foo';
+        $_SERVER['HTTP_HOST'] = 'localhost';
+
+        if (!\defined('TL_FILES_URL')) {
+            \define('TL_FILES_URL', '');
+        }
```

Comment:
This seems wrong. Why is it necessary?

Reply:
Otherwise `Controller::setStaticUrls()` wants to set them:

https://github.com/contao/contao/blob/724fd43e33dc8bcaa720a71142bb521688a1c71b/core-bundle/src/Resources/contao/library/Contao/Controller.php#L1881-L1904

In which case you need more setup (files context, …) but the constant will be defined no matter what.